pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,444 @@
+
+	cb.objectSizeRequests.Add(1)
+
+	hits := cb.cache.Fetch(ctx, []string{key})
+	if s := hits[key]; len(s) == 8 {
+		cb.objectSizeHits.Add(1)
+		return binary.BigEndian.Uint64(s), nil
+	}
+
+	size, err := cb.bucket.ObjectSize(ctx, name)
+	if err != nil {
+		return 0, err
+	}
+
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], size)
+	cb.cache.Store(ctx, map[string][]byte{key: buf[:]}, ttl)
+
+	return size, nil
+}
+
+func (cb *CachingBucket) getRangeChunkFile(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	cb.requestedChunkBytes.Add(float64(length))
+
+	size, err := cb.cachedObjectSize(ctx, name, cb.config.ChunkObjectSizeTTL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get size of chunk file: %s", name)
+	}
+
+	// If length goes over object size, adjust length. We use it later to limit number of read bytes.
+	if uint64(offset+length) > size {
+		length = int64(size - uint64(offset))
+	}
+
+	// Start and end range are block-aligned offsets into object, that we're going to read.
+	startRange := (offset / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	endRange := ((offset + length) / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	if (offset+length)%cb.config.ChunkBlockSize > 0 {
+		endRange += cb.config.ChunkBlockSize
+	}
+
+	// The very last block in the object may have length that is not divisible by block size.
+	lastBlockOffset := endRange - cb.config.ChunkBlockSize
+	lastBlockLength := int(cb.config.ChunkBlockSize)
+	if uint64(endRange) > size {
+		lastBlockOffset = (int64(size) / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+		lastBlockLength = int(int64(size) - lastBlockOffset)
+	}
+
+	numBlocks := (endRange - startRange) / cb.config.ChunkBlockSize
+
+	offsetKeys := make(map[int64]string, numBlocks)
+	keys := make([]string, 0, numBlocks)
+
+	for off := startRange; off < endRange; off += cb.config.ChunkBlockSize {
+		end := off + cb.config.ChunkBlockSize
+		if end > int64(size) {
+			end = int64(size)
+		}
+
+		k := cachingKeyObjectBlock(name, off, end)
+		keys = append(keys, k)
+		offsetKeys[off] = k
+	}
+
+	// Try to get all blocks from the cache.
+	hits := cb.cache.Fetch(ctx, keys)
+	for _, b := range hits {
+		cb.fetchedChunkBytes.WithLabelValues(originCache).Add(float64(len(b)))
+	}
+
+	if len(hits) < len(keys) {
+		if hits == nil {
+			hits = map[string][]byte{}
+		}
+
+		err := cb.fetchMissingChunkBlocks(ctx, name, startRange, endRange, offsetKeys, hits, lastBlockOffset, lastBlockLength)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return newBlocksReader(cb.config.ChunkBlockSize, offsetKeys, hits, offset, length), nil
+}
+
+type rng struct {
+	start, end int64
+}
+
+// Fetches missing blocks and stores them into "hits" map.
+func (cb *CachingBucket) fetchMissingChunkBlocks(ctx context.Context, name string, startRange, endRange int64, cacheKeys map[int64]string, hits map[string][]byte, lastBlockOffset int64, lastBlockLength int) error {
```

Comment:
even if this is not exported it's really nice if comments starts with method name.